implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/initiator_state_machine.rs

```diff
@@ -0,0 +1,115 @@
+use crate::secure_channel::handshake::handshake::Handshake;
+use crate::secure_channel::handshake::handshake_state_machine::{
+    Action, CommonStateMachine, Event, HandshakeKeys, HandshakeResults, IdentityAndCredentials,
+    StateMachine, Status,
+};
+use crate::{Credential, Identities, IdentityIdentifier, Role, TrustContext, TrustPolicy, XXVault};
+use delegate::delegate;
+use ockam_core::async_trait;
+use ockam_core::compat::sync::Arc;
+use ockam_core::compat::{boxed::Box, vec::Vec};
+use ockam_core::errcode::{Kind, Origin};
+use ockam_core::{Error, Result};
+use ockam_vault::PublicKey;
+use Action::*;
+use Event::*;
+use Role::*;
+use Status::*;
+
+/// Implementation of a state machine for the key exchange on the initiator side
+#[async_trait]
+impl StateMachine for InitiatorStateMachine {
+    async fn on_event(&mut self, event: Event) -> Result<Action> {
+        let state = self.handshake.state.clone();
+        match (state.status, event) {
+            // Initialize the handshake and send message 1
+            (Initial, Initialize) => {
+                self.initialize_handshake().await?;
+                let message1 = self.encode_message1().await?;
+
+                // Send message 1 and wait for message 2
+                self.handshake.state.status = WaitingForMessage2;
+                Ok(SendMessage(message1))
+            }
+            // Process message 2 and send message 3
+            (WaitingForMessage2, ReceivedMessage(message)) => {
+                let message2_payload = self.decode_message2(message).await?;
+                let their_identity_payload =
+                    CommonStateMachine::deserialize_payload(message2_payload)?;
+                self.verify_identity(their_identity_payload, &self.handshake.state.rs()?)
+                    .await?;
+                let message3 = self.encode_message3(self.identity_payload.clone()).await?;
+                self.set_final_state(Initiator).await?;
+                Ok(SendMessage(message3))
+            }
+            // incorrect state / event
+            (s, e) => Err(Error::new(
+                Origin::Channel,
+                Kind::Invalid,
+                format!(
+                    "Unexpected combination of initiator state and event {:?}/{:?}",
+                    s, e
+                ),
+            )),
+        }
+    }
+
+    fn get_handshake_results(&self) -> Option<HandshakeResults> {
+        self.make_handshake_results(self.get_handshake_keys())
+    }
+}
+
+/// Implementation of the state machine actions, delegated to the Handshake module
+pub(super) struct InitiatorStateMachine {
+    pub(super) common: CommonStateMachine,
+    pub(super) handshake: Handshake,
+    pub(super) identity_payload: Vec<u8>,
```

Comment:
It also includes credentials, right? Should we describe that in the name somehow? Or maybe in a doc-comment?

Reply:
Also includes the static key signature I guess